op-node/rollup/derive/channel_frame.go

```diff
@@ -0,0 +1,134 @@
+package derive
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+)
+
+var ErrNotEnoughFrameBytes = errors.New("not enough available bytes for the frame")
+
+// Data Format
+//
+// frame = channel_id ++ frame_number ++ frame_data_length ++ frame_data ++ is_last
+//
+// channel_id        = random ++ timestamp
+// random            = bytes32
+// timestamp         = uvarint
+// frame_number      = uvarint
+// frame_data_length = uvarint
+// frame_data        = bytes
+// is_last           = bool
+
+type Frame struct {
+	Id          ChannelID
+	FrameNumber uint64
+	Data        []byte
+	IsLast      bool
+}
+
+// MarshalBinary writes the frame to `w`.
+// It returns the number of bytes written as well as any
+// error encountered while writing.
+func (f *Frame) MarshalBinary(w io.Writer) (int, error) {
+	n, err := w.Write(f.Id.Data[:])
+	if err != nil {
+		return n, err
+	}
+	l, err := w.Write(makeUVarint(f.Id.Time))
+	n += l
+	if err != nil {
+		return n, err
+	}
+	l, err = w.Write(makeUVarint(f.FrameNumber))
+	n += l
+	if err != nil {
+		return n, err
+	}
+
+	l, err = w.Write(makeUVarint(uint64(len(f.Data))))
+	n += l
+	if err != nil {
+		return n, err
+	}
+	l, err = w.Write(f.Data)
+	n += l
+	if err != nil {
+		return n, err
+	}
+	if f.IsLast {
+		l, err = w.Write([]byte{1})
+		n += l
+		if err != nil {
+			return n, err
+		}
+	} else {
+		l, err = w.Write([]byte{0})
+		n += l
+		if err != nil {
+			return n, err
+		}
+	}
+	return n, nil
+}
+
+// TODO: Is there standard interfaced for this
+// or should I just use a concrete struct in frame.Unmarshall?
+type ByteAndNormalReader interface {
+	io.Reader
+	io.ByteReader
+}
+
+// UnmarshalBinary consumes a full frame from the reader.
+// If `r` fails a read, it returns the error from the reader
+// as well as the total number of read bytes.
+// If there are not enough bytes available ErrNotEnoughFrameBytes
+// is returned along with the number of bytes read.
+func (f *Frame) UnmarshalBinary(r ByteAndNormalReader) error {
+	_, err := io.ReadFull(r, f.Id.Data[:])
+	if err != nil {
+		return fmt.Errorf("error reading Id: %w", err)
+	}
+	f.Id.Time, err = binary.ReadUvarint(r)
+	if err != nil {
+		return fmt.Errorf("error reading Id time: %w", err)
+	}
+	// stop reading and ignore remaining data if we encounter a zeroed ID
+	if f.Id == (ChannelID{}) {
+		return nil
```

Comment:
```suggestion
		return io.EOF
```
We can stop reading the input reader completely at this point, instead of allowing the channel bank to read a next frame after this.